RenderMan `UsdPreviewSurface`

ericmehl opened the pull request:
This add the basics for using a `UsdPreviewSurface` with `IECoreRenderMan`. It supports `UsdPreviewSurface` itself, as well as `UsdPrimvarReader` nodes.

Note : At least on Windows, you need to  `import IECoreRenderMan` before trying to render with RenderMan. I think that's because we don't use anything from `IECoreRenderMan.dll` in `GafferRenderMan` yet, but I haven't looked into it fully.

There are some hints from https://github.com/PixarAnimationStudios/OpenUSD/blob/dev/third_party/renderman-26/plugin/hdPrman/matfiltConvertPreviewMaterial.cpp that informed the conversion to `PxrSurface`, such as for changing the specular model, that may be useful for review and future additions.

This replaces #6299. This PR is made from a branch on GafferHQ rather than my own repository so it can build and test RenderMan.

### Checklist ###

- [X] I have read the [contribution guidelines](https://github.com/GafferHQ/gaffer/blob/main/CONTRIBUTING.md).
- [X] I have updated the documentation, if applicable.
- [X] I have tested my change(s) in the test suite, and added new test cases where necessary.
- [X] My code follows the Gaffer project's prevailing coding style and conventions.

murraystevenson commented on shaders/__usd/__UsdPreviewSurfaceParameters.osl:
```diff
@@ -0,0 +1,196 @@
+        [[ string widget = "null" ]],
+)
+{
+    // To achieve double-sided materials from a geometry-level property,
+    // we set a special user:hydra:doubleSided attribute during renderParam
+    // setup (see hdPrman/renderParam.cpp) and read that here. We also check 
+    // whether we are shading the front or back based on backfacing().
+    // 
+    // NB: We have to do this test in the shader because the *DoubleSided
+    // parameters of PxrSurface are NOT CONNECTABLE, so we cannot just switch
+    // them based on an output from this shader. We have to make the backface-
+    // shading decision here.
+    
+    int isDoubleSided = 0;
+    getattribute("user:hydra:doubleSided", isDoubleSided);
```
We'll need to adapt this to be more Gaffer-specific, changing the `isDoubleSided` fallback to `1` to match Gaffer's default behaviour and renaming `user:hydra:doubleSided`. We'd also need to shuffle Gaffer's `doubleSided` attribute to that new name, either in a render adaptor or in the RenderMan backend itself, maybe @johnhaddon has some thoughts here...

ericmehl replied:
Relating what we spoke about offline, we decided to keep using the `user:hydra:doubleSided` attribute as it draws attention to the somewhat unusual nature of how `doubleSided` is handled with the `PxrSurface` shader. I'm shuffling `doubleSided` into that parameter in https://github.com/GafferHQ/gaffer/pull/6302/commits/474584a608bdf68e853391bc7ef0f73ffd3b78d0, as well as setting the `isDoubleSided` default to `1` to match Gaffer's default.

ericmehl replied:
We ended up deciding to _not_ shuffle into the `user:hydra:doubleSided` parameter and instead only default the double sided variable in `__UsdPreviewSurfaceParameters" to `1`. The `UsdPreviewSurfaceParameters` shader does not do true double sided - it renders the back side black when double sided is off, rather than transparent.

Since you can't get correct non-double sided behavior and it isn't clear which non-standard behavior to go with, we are opting for the less complex option of always using double sided USDPreviewSurface.

This is done in https://github.com/GafferHQ/gaffer/pull/6302/commits/f664dd94ba42cba3088c4286257e9566a3c85d5d

murraystevenson reviewed (commented):
Thanks Eric! It looks like there's almost all of the required functionality here to cover the UsdPreviewSurface dependent bits of the render pass adaptors. The missing piece is the handling of double-sided geometry in `__UsdPreviewSurfaceParameters.osl`, I've left a comment inline with some thoughts...

murraystevenson commented:
Thanks Eric, it'd be worth getting this squashed and rebased ready for merging. With some additional GafferRenderMan PRs merged in the meantime, I suspect we'll also need to update [this](https://github.com/GafferHQ/gaffer/blob/bf3392d711f71c4db81237065ef9b71e9c89fcff/src/IECoreRenderMan/Globals.cpp#L40) new reference to `"ShaderNetworkAlgo.h"`.

ericmehl commented:
> I suspect we'll also need to update [this](https://github.com/GafferHQ/gaffer/blob/bf3392d711f71c4db81237065ef9b71e9c89fcff/src/IECoreRenderMan/Globals.cpp#L40) new reference to `"ShaderNetworkAlgo.h"`.

Yes, that was needed but I think that should be the only change. I've rebased and squashed the fixups down. Hopefully ready for a merge.
